src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -166,19 +166,24 @@ internal ThreadPoolBoundHandle RequestQueueBoundHandle
         {
             get
             {
-                if (_requestQueueBoundHandle == null)
+                ThreadPoolBoundHandle handle = Volatile.Read(ref _requestQueueBoundHandle);
+
+                if (handle != null)
                 {
-                    lock (_internalLock)
+                    return handle;
+                }
+
+                lock (_internalLock)
+                {
+                    handle = _requestQueueBoundHandle;
+                    if (handle == null)
                     {
-                        if (_requestQueueBoundHandle == null)
-                        {
-                            _requestQueueBoundHandle = ThreadPoolBoundHandle.BindHandle(_requestQueueHandle);
-                            if (NetEventSource.IsEnabled) NetEventSource.Info($"ThreadPoolBoundHandle.BindHandle({_requestQueueHandle}) -> {_requestQueueBoundHandle}");
-                        }
+                        _requestQueueBoundHandle = handle = ThreadPoolBoundHandle.BindHandle(_requestQueueHandle);
```

Comment:
Can the updated DisconnectAsyncResult ctor ever run after CloseRequestQueueHandle completes? I'm not familiar with the HttpListener code, but it looks like RegisterForDisconnectNotification isn't synchronized.

Could this risk BindHandle throwing an ArgumentException since _requestQueueHandle would be closed at that point?